File: 백준/Silver/20529. 가장 가까운 세 사람의 심리적 거리/가장 가까운 세 사람의 심리적 거리.py
from sys import stdin

input = stdin.readline

T = int(input())

MBTI_type = ('ISTJ', 'ISFJ', 'INFJ', 'INTJ', 'ISTP', 'ISFP', 'INFP', 'INTP', 'ESTP', 'ESFP', 'ENFP', 'ENTP', 'ESTJ', 'ESFJ', 'ENFJ', 'ENTJ')


def diff_MBTI(mbti1, mbti2):
    ret = 0
    
    for i in range(4):
        if mbti1[i] == mbti2[i]:
            continue
        ret += 1
        
    return ret


for _ in range(T):
    n = int(input())
    MBTI = list(input().split())

    MBTI_map = {}

    for data in MBTI:
        if MBTI_map.get(data) is None:
            MBTI_map[data] = 1
            continue
        
        MBTI_map[data] += 1


    total = 100
    mbti_count = dict.fromkeys(MBTI_type, 0)
    
    for mbti1 in MBTI_type:
        if MBTI_map.get(mbti1) is None:
            continue
        
        mbti_count[mbti1] += 1
        
        for mbti2 in MBTI_type:
            if MBTI_map.get(mbti2) is None or MBTI_map.get(mbti2) < (mbti_count[mbti2] + 1):
                continue

            mbti_count[mbti2] += 1
            
            for mbti3 in MBTI_type:
                if MBTI_map.get(mbti3) is None or MBTI_map.get(mbti3) < (mbti_count[mbti3] + 1):
                    continue

                total = min(total, diff_MBTI(mbti1, mbti2) + diff_MBTI(mbti2, mbti3) + diff_MBTI(mbti1, mbti3))
                
            mbti_count[mbti2] -= 1
            
        mbti_count[mbti1] -= 1

    print(total)

File: 백준/Silver/20529. 가장 가까운 세 사람의 심리적 거리/가장 가까운 세 사람의 심리적 거리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.StringTokenizer;

class Main {
    private static final String[] mbtiType = {"ISTJ", "ISFJ", "INFJ", "INTJ", "ISTP", "ISFP",
    "INFP", "INTP", "ESTP", "ESFP", "ENFP", "ENTP", "ESTJ", "ESFJ", "ENFJ", "ENTJ"};

    private static HashMap<String, Integer> mbti;
    private static int distTotal;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));


        int T = Integer.parseInt(br.readLine());

        StringBuilder answer = new StringBuilder();

        while(T-- > 0){
            int n = Integer.parseInt(br.readLine());

            StringTokenizer tokenizer = new StringTokenizer(br.readLine());

            initMbti();

            for(int i=0 ; i<n ; i++){
                String type = tokenizer.nextToken();
                mbti.put(type, mbti.get(type) + 1);
            }

            distTotal = Integer.MAX_VALUE;

            dfs(new ArrayDeque<>());

            answer.append(distTotal).append("\n");
        }

        System.out.print(answer);
    }

    private static void initMbti(){
        mbti = new HashMap<>();

        for(String type : mbtiType){
            mbti.put(type, 0);
        }
    }

    private static void dfs(ArrayDeque<String> path){
        if(path.size() == 3){
            String[] select = new String[3];
            int idx = 0;

            for(String type : path){
                select[idx++] = type;
            }

            distTotal = Math.min(distTotal, calculateDistance(select[0], select[1]) + calculateDistance(select[1], select[2]) + calculateDistance(select[0], select[2]));
            return;
        }

        for(int i=0 ; i<16 ; i++){
            String type = mbtiType[i];
            if(mbti.get(type) == 0)
                continue;

            mbti.put(type, mbti.get(type) - 1);
            path.add(mbtiType[i]);
            dfs(path);
            path.pollLast();
            mbti.put(type, mbti.get(type) + 1);
        }
    }

    private static int calculateDistance(String stu1, String stu2){
        int dist = 0;

        for(int i=0 ; i<4 ; i++){
            if(stu1.charAt(i) == stu2.charAt(i))
                continue;

            dist++;
        }

        return dist;
    }
}